cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -1241,6 +1373,45 @@ std::shared_ptr<ScalarFunction> MakeUnaryArithmeticFunctionNotNull(
   return func;
 }
 
+// Like MakeUnaryArithmeticFunction, but for unary rounding functions that control
+// kernel dispatch based on RoundMode, only on non-null output.
+template <template <RoundMode RndMode> class Op, typename Opts>
+std::shared_ptr<ScalarFunction> MakeUnaryRoundFunction(
+    std::string name, const FunctionDoc* doc,
+    const FunctionOptions* default_options = NULLPTR, KernelInit init = NULLPTR) {
+  auto func = std::make_shared<ArithmeticFloatingPointFunction>(name, Arity::Unary(), doc,
+                                                                default_options);
+  for (const auto& ty : FloatingPointTypes()) {
+    // Order of ArrayKernelExec w.r.t. RoundMode needs to follow the order of
+    // values in RoundMode definition (see api_scalar.h) because they are used as
+    // indexing values.
+    std::vector<ArrayKernelExec> execs = {
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::TOWARDS_NEG_INFINITY>>(
+            ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::TOWARDS_POS_INFINITY>>(
+            ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::TOWARDS_ZERO>>(ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::TOWARDS_INFINITY>>(ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::HALF_NEG_INFINITY>>(
+            ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::HALF_POS_INFINITY>>(
+            ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::HALF_TOWARDS_ZERO>>(
+            ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary,
+                                        Op<RoundMode::HALF_TOWARDS_INFINITY>>(ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::HALF_TO_EVEN>>(ty),
+        GenerateArithmeticFloatingPoint<ScalarUnary, Op<RoundMode::HALF_TO_ODD>>(ty),
+    };
```

Comment:
I wonder, given the branch predictor, if a switch-case might perform similarly to generating separate kernels for every type and rounding mode, if you're concerned about having to place the rounding mode in the template parameters.

Reply:
A switch-case was the initial approach taken but after code reviews, the conclusion was to resolve the kernel during dispatching rather than performing the check internally for each value/invocation. I am not overtly concerned about the explicit rounding modes in template parameters, but the `vector` order and `RoundMode`s need to match.

Reply:
Ah whoops, thanks for the context.